common/time-service/src/mock.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Sleep, SleepTrait, TimeServiceTrait, ZERO_DURATION};
+use diem_infallible::Mutex;
+use futures::future::Future;
+use std::{
+    cmp::max,
+    collections::hash_map::{DefaultHasher, HashMap},
+    fmt::Debug,
+    hash::BuildHasherDefault,
+    pin::Pin,
+    sync::{Arc, MutexGuard},
+    task::{Context, Poll, Waker},
+    time::Duration,
+};
+
+// TODO(philiphayes): use more robust solution that'll let us control the seed.
+type DeterministicHasher = BuildHasherDefault<DefaultHasher>;
+
+type SleepIndex = usize;
+
+/// A [`TimeService`] that simulates time and allows for fine-grained control over
+/// advancing time and waking up sleeping tasks.
+#[derive(Clone, Debug)]
+pub struct MockTimeService {
+    inner: Arc<Mutex<Inner>>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    /// The current simulated time.
+    now: Duration,
+    /// The next free index in the `pending` queue that we can allocate to a new waiter.
+    next_sleep_index: SleepIndex,
+    /// A queue of pending `SleepEntry`s, each with a deadline for when we should
+    /// wake them.
+    ///
+    /// This `HashMap` also uses a hasher with constant seed (in the future the
+    /// seed should be configurable) to help improve test determinism.
+    pending: HashMap<SleepIndex, SleepEntry, DeterministicHasher>,
+}
+
+/// A timer entry owned by the `MockTimeService` in the `pending` queue.
+///
+/// These are indexed by `SleepIndex`, which you get when you `register_sleep` a
+/// new waiter.
+#[derive(Debug)]
+struct SleepEntry {
+    /// The time when this sleep task expires and should be woken up.
+    deadline: Duration,
+    /// When the corresponding `MockSleep` future polls and isn't expired
+    /// yet, it will register its `Waker` here for the `MockTimeService`
+    maybe_waker: Option<Waker>,
+}
+
+/// A [`Future`] that resolves when the simulated time in the [`MockTimeService`]
+/// advances past the deadline. When these are pending, they have a `SleepEntry`
+/// in the `MockTimeService`'s `pending` queue.
+#[derive(Debug)]
+pub struct MockSleep {
+    /// A handle to the `MockTimeService` so we can get the underlying `SleepEntry`.
+    time_service: MockTimeService,
+    /// The index of this `MockSleep`'s `SleepEntry` in the `MockTimeService`'s
+    /// `pending` queue. If the `pending` queue doesn't contain our entry, then
+    /// this sleep must be completed.
+    entry_index: SleepIndex,
+}
+
+/////////////////////
+// MockTimeService //
+/////////////////////
+
+impl MockTimeService {
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(Inner {
+                now: ZERO_DURATION,
+                next_sleep_index: 0,
+                pending: HashMap::default(),
```

Comment:
BTreeMap is probably more performant here or a BinaryTree either as a min heap -- from experience playing with simulators...

Reply:
Right, I think I can do `BTreepMap<(Duration, SleepIndex), Option<Waker>>` instead of this `HashMap<SleepIndex, SleepEntry>`.